File: C++/SubmodularFunction/SubmodularFunction/ConnectedDetachment.cpp
#include "Submodular.h"


using namespace OnigiriSubmodular;

void ConnectedDetachment::SetVariables(int n, double* modular,map<int,bool>* edges){
	map<int,bool>::iterator itr;
	ConnectedDetachment::n = n;
	ConnectedDetachment::modular = new double[n];
	ConnectedDetachment::edges = new map<int,bool>[n];
	for(int i=0;i<n;i++){
		ConnectedDetachment::modular[i] = modular[i];
		ConnectedDetachment::edges[i] = *new map<int,bool>();
		for(itr = edges[i].begin();itr!=edges[i].end();itr++)
		{
			int u = itr->first;
			ConnectedDetachment::edges[i][u] = true;
		}//foreach u
	}
	ConnectedDetachment::fOfEmpty = 0;

	uf =new UnionFind(n);
    used=new bool[n];
    numComplement=new int[n];
}

ConnectedDetachment::ConnectedDetachment(int n,double* modular,map<int,bool>* edges){
	SetVariables(n,modular,edges);
}

ConnectedDetachment::ConnectedDetachment(string path){ 
	ifstream file(path);
	int n;
	if(file.fail()) {
		cerr << path + " does not exist."<<endl;
		exit(0);
	}

	file>>n;
	double* modular =new double[n];
	for(int i=0;i<n;i++){
		file>>modular[i];
	}

	map<int,bool>* edges = new map<int,bool>[n];
	for(int i=0;i<n;i++){
		int cnt;
		file>>cnt;
		edges[i] = map<int,bool>();
		for(int j=0;j<cnt;j++){
			int cur;
			file>>cur;
			edges[i][cur] = true;
		}
	}
	file.close();
	SetVariables(n,modular,edges);
	delete[]modular;
	for(int i=0;i<n;i++){
		edges[i].clear();
	}
	delete[] edges;
}

ConnectedDetachment::~ConnectedDetachment(){
	delete[]ConnectedDetachment::modular;
	for(int i=0;i<n;i++){
		edges[i].clear();
	}
	delete[] ConnectedDetachment::edges;

	delete uf;
	delete[]used;
	delete[]numComplement;
}


double ConnectedDetachment::Value(const int* order,int cardinality){
	double res = 1;
	for (int i = 0; i < cardinality; i++)
	{
		res -= modular[order[i]];
	}//for i
	res += CountIncidentEdge(order, cardinality);
	res -= CountComponentOfComplementGraph(order, cardinality);
	return res;
}


void ConnectedDetachment::Base(const int* order,double* base){
	map<int,bool>::iterator itr;
	SetArrayOfComplementConnectedComponent(order);
	SetUsed();
	for (int i = 0; i < n; i++)
	{
		int v = order[i];
		double cur = 0;
		cur -= modular[v];

		for(itr = edges[v].begin();itr!=edges[v].end();itr++)
		{
			if (!used[itr->first])
			{
				cur++;
			}//if
		}//foreach u
		cur -= numComplement[v] - (i == 0 ? 1 : numComplement[order[i - 1]]);
		used[v] = true;
		base[v] = cur;
	}//for i
}

void ConnectedDetachment::SetArrayOfComplementConnectedComponent(const int*order)
{
	map<int,bool>::iterator itr;
	SetUsed();
	uf->Clear();
	int component = 0;
	for(int i=0;i<n;i++){
		numComplement[i] = 0;
	}
	for (int i = n - 1; i > 0; i--)
	{
		component++;
		int v = order[i];
		for(itr = edges[v].begin();itr!=edges[v].end();itr++)
		{
			int u = itr->first;
			if (used[u]&&!uf->Same(v,u))
			{
				component--;
				uf->Unite(v, u);
			}//if
		}//foreach u
		used[v] = true;
		numComplement[order[i-1]] = component;
	}//forrev i
}

int ConnectedDetachment::CountIncidentEdge(const int* order, int cardinality)
{
	map<int,bool>::iterator itr;
	SetUsed();
	int res = 0;
	for (int i = 0; i < cardinality; i++)
	{
		int v = order[i];
		for(itr = edges[v].begin();itr!=edges[v].end();itr++)
		{
			if (!used[ itr->first])
			{
				res++;
			}//if
		}//foreach u
		used[v] = true;
	}//for v
	return res;
}

int ConnectedDetachment::CountComponentOfComplementGraph(const int* order, int cardinality)
{
	map<int,bool>::iterator itr;
	uf->Clear();
	int res = n - cardinality;
	SetUsed();
	for (int i = cardinality; i < n; i++)
	{
		used[order[i]] = true;
	}//for i
	for (int i = cardinality; i < n; i++)
	{
		int v = order[i];
		for(itr = edges[v].begin();itr!=edges[v].end();itr++)
		{
			int u = itr->first;
			if (used[u]&&!uf->Same(order[i],u))
			{
				uf->Unite(order[i], u);
				res--;
			}//if
		}//foreach u
	}//for i
	return res;
}

void ConnectedDetachment::SetUsed(){
	for(int i=0;i<n;i++){
		used[i] = false;
	}
}


#if _DEBUG

void ConnectedDetachment::TestCalcBase(const int* order){
	double* b0 = new double[n];
	double* b1 = new double[n];
	ConnectedDetachment::CalcBase(order,b0);
	SubmodularOracle::CalcBase(order,b1);
	bool same  = true;
	for(int i=0;i<n;i++){
		same&=(b0[i]==b1[i]);
	}
	delete[]b0;
	delete[]b1;
	if (!same)
	{
		throw new exception();
	}//if
}
#endif

File: C#/SubmodularFunction/Submodular/ConnectedDetachment.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.IO;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Onigiri.Submodular
{
    public class ConnectedDetachment:SubmodularOracle
    {
        double[] modular;
        HashSet<int>[] edges;
        UnionFind uf;   //for calculation
        bool[] used;    //for calculation
        int[] numComilement;    //for calculation

        public ConnectedDetachment(int n, double[] modular, HashSet<int>[] edges)
        {
            SetVariables(n, modular, edges);
        }

        public ConnectedDetachment(string path)
        {
            var streamReader = new StreamReader(path);
            int n = int.Parse(streamReader.ReadLine());
            var modular = new double[n];
            for (int i = 0; i < n; i++)
            {
                modular[i] = double.Parse(streamReader.ReadLine());
            }//for i
            var edges = new HashSet<int>[n];
            for (int i = 0; i < n; i++)
            {
                edges[i] = new HashSet<int>();
                var list = Array.ConvertAll<string, int>(streamReader.ReadLine().Split(new char[] { ' ' }, StringSplitOptions.RemoveEmptyEntries), int.Parse);
                foreach (var e in list.Skip(1))
                {
                    edges[i].Add(e);
                }//foreach e
            }//for i
            streamReader.Close();
            SetVariables(n, modular, edges);
        }

        private void SetVariables(int n, double[] modular, HashSet<int>[] edges)
        {
            SetVariable(n, 0);
            this.modular = modular;
            this.edges = edges;

            uf = new UnionFind(N);
            used = new bool[N];
            numComilement = new int[N];
        }

        internal override double Value(int[] order, int cardinality)
        {
            double res = 1;
            for (int i = 0; i < cardinality; i++)
            {
                res -= modular[order[i]];
            }//for i
            res += CountIncidentEdge(order, cardinality);
            res -= CountComponentOfComilementGraih(order, cardinality);
            return res;
        }

        internal override void Base(int[] order, double[] b)
        {
            SetArrayOfComilementConnectedComponent(order);
            SetUsed();
            for (int i = 0; i < N; i++)
            {
                int v = order[i];
                double cur = 0;
                cur -= modular[v];

                foreach (var u in edges[v])
                {
                    if (!used[u])
                    {
                        cur++;
                    }//if
                }//foreach u
                cur -= numComilement[v] - (i == 0 ? 1 : numComilement[order[i - 1]]);
                used[v] = true;
                b[v] = cur;
            }//for i
        }

        private void SetArrayOfComilementConnectedComponent(int[]order)
        {
            for (int i = 0; i < N; i++)
            {
                numComilement[i] = 0;
            }//for i
            uf.Clear();
            SetUsed();
            int component = 0;
            for (int i = N - 1; i > 0; i--)
            {
                component++;
                int v = order[i];
                foreach (var u in edges[v])
                {
                    if (used[u]&&!uf.Same(v,u))
                    {
                        component--;
                        uf.Unite(v, u);
                    }//if
                }//foreach u
                used[v] = true;
                numComilement[order[i-1]] = component;
            }//forrev i
        }

        private int CountIncidentEdge(int[] order, int cardinality)
        {
            SetUsed();
            int res = 0;
            for (int i = 0; i < cardinality; i++)
            {
                int v = order[i];
                foreach (var u in edges[v])
                {
                    if (!used[u])
                    {
                        res++;
                    }//if
                }//foreach u
                used[v] = true;
            }//for v
            return res;
        }

        private int CountComponentOfComilementGraih(int[] order, int cardinality)
        {
            uf.Clear();
            int res = N - cardinality;
            SetUsed();
            for (int i = cardinality; i < N; i++)
            {
                used[order[i]] = true;
            }//for i
            for (int i = cardinality; i < N; i++)
            {
                foreach (var u in edges[order[i]])
                {
                    if (used[u]&&!uf.Same(order[i],u))
                    {
                        uf.Unite(order[i], u);
                        res--;
                    }//if
                }//foreach u
            }//for i
            return res;
        }

        private void SetUsed()
        {
            for (int i = 0; i < N; i++)
            {
                used[i] = false;
            }//for i
        }

        //protected override void ContractDerived(int element)
        //{
        //    throw new NotImilementedException();
        //}
   

    }
}
